### ISSUE TITLE
Single Tap twice as slow in Appium 1.15.1

### ISSUE THREAD

**Issue by sshock** [appium/appium-uiautomator2-server#365]
Sometime between Appium 1.14.2 and 1.15.1 (between uiautomator2 3.7.0 and 4.3.0), a slowdown occurred in single tapping keys on the virtual keyboard with TouchAction.  It takes about twice as long in the newer version.

Here is a screen clip using Appium 1.9.1 (fast):
https://drive.google.com/file/d/1ESVnsraUsEsuDM-O6nLJfD4jzZICsaQH/view?usp=sharing

Here is a screen clip using Appium 1.17.1 (slow):
https://drive.google.com/file/d/1cBISOIwys4aP4R8DZdxl2S8J6xJPcifJ/view?usp=sharing

Job Times (3 runs)
---------------
Appium Version | UIAutomator2 version | time (seconds)
--- | --- | ---
1.9.1 | 1.18.0 | 66, 89, 63
1.11.0 | 3.1.1 | 66, 74, 70
1.13.0 | 3.5.1 | 59, 58, 60
1.14.2 | 3.7.0 | 69, 70, 58
1.15.1 | 4.3.0 | 143, 142, 131

Those tests were all performed with the exact same code, written in ruby using the ruby appium_lib.  The code works by making multiple calls to a `single_tap` method that just calls `::Appium::TouchAction.new(driver).tap(:x => ###, :y => ###)` with the appropriate x and y coordinates for the key to be entered.

Here is the implementation of our single_tap and related methods it uses if it helps:

```ruby
 def single_tap(target, safe: false)
    opts = xy_coordinates(gaussian_coordinates_in_bounds(bounds(target)))
    touch_action.tap(opts).perform
  end

  def bounds(target)
    if target.is_a?(::Hash)
      target
    else
      element_bounds(target)
    end
  end

  def element_bounds(element)
    {
      :x_min => element.rect.x,
      :y_min => element.rect.y,
      :x_max => element.rect.x + element.rect.width,
      :y_max => element.rect.y + element.rect.height,
    }
  end

  def gaussian_coordinates_in_bounds(bounds)
    # bounds (hash): :x_min => int, :y_min => int, :x_max => int, :y_max => int
    x_coord = gaussian_by_bounds(bounds[:x_min], bounds[:x_max])
    y_coord = gaussian_by_bounds(bounds[:y_min], bounds[:y_max])
    [x_coord, y_coord]
  end

  def xy_coordinates(coordinates)
    [:x, :y].zip(coordinates).to_h
  end

  def touch_action
    ::Appium::TouchAction.new(driver)
  end
```

I don't know if the issue may be related to waitForIdle, but it seems like it might be since the slowdown appears to go away if I set waitForIdleTimeout to 0, but unfortunately that doesn't actually work because although you can see it "tapping" all the keys, many of them don't register (i.e., the characters don't actually get entered into the edit box).

~If you need me to, I can do more tests to verify exactly which version of uiautomator2 this slowdown was introduced in, but a good guess would be in uiautomator2 2.6.0, since that affected the AX tree refresh and it is where other slowdowns were noticed as demonstrated in #351.~  I did more testing and have narrowed down the slowdown to sometime between Appium 1.14.2 and 1.15.1.  I can narrow it down further if you need me to.

**Comment by sshock** [comment]
I tried the following change, but it didn't help:

```patch
diff --git a/app/scrapes/recipe_helpers/scene.rb b/app/scrapes/recipe_helpers/scene.rb
index fca513a4..2cc72a90 100644
--- a/app/scrapes/recipe_helpers/scene.rb
+++ b/app/scrapes/recipe_helpers/scene.rb
@@ -143,11 +143,12 @@ private
   end

   def element_bounds(element)
+    r = element.rect
     {
-      :x_min => element.rect.x,
-      :y_min => element.rect.y,
-      :x_max => element.rect.x + element.rect.width,
-      :y_max => element.rect.y + element.rect.height,
+      :x_min => r.x,
+      :y_min => r.y,
+      :x_max => r.x + r.width,
+      :y_max => r.y + r.height,
     }
   end

@@ -184,7 +185,7 @@ private
   end

   def touch_action
-    ::Appium::TouchAction.new(driver)
+    @touch_action ||= ::Appium::TouchAction.new(driver)
   end

   def xy_coordinates(coordinates)
```

**Comment by mykola-mokhnach** [comment]
Try to disable `trackScrollEvents` in settings. Also, the most time-precise actions implementation are W3C actions in the recent master.

**Comment by sshock** [comment]
Unfortunately, setting `trackScrollEvents` to false had no effect.  But what's really weird about that is I noticed that `driver.get_settings` shows it as already being false, even before I modified it with update_setting, but that doesn't make any sense because it should be true by default.

Thanks for the other idea too of using W3C actions; I may give that a try later, but I really want to get to the bottom of this slowdown, especially if there is any chance whatever is going on is affecting more than just touch actions.

**Comment by sshock** [comment]
I've narrowed it down further.  The slowdown occurred in uiautomator2 v4.0.0.  The last fast one was v3.8.1.

Which means the slowdown was introduced in this commit:
752a8a8 - feat: Switch UiAutomator2 server to W3C protocol

Any ideas why that change would slow down TouchAction single tapping?  If there is any part of that change you would like me to tweak and test I can.

**Comment by sshock** [comment]
@mykola-mokhnach I put in some debug logs to try and determine for sure if `trackScrollEvents` is true or false, but it turns out that my code does not even call into the `io.appium.uiautomator2.core.InteractionController` at all, which is the class that performs various touch actions and checks that setting.

So..., I still have no answer for why that setting appears to be false even before setting it to false, but I don't think it matters in my case since I'm not even using that code.

Turns out my code is hitting the `tap` endpoint which calls into the `io.appium.uiautomator2.handler.Click` class, which in turn calls into `getUiDevice().click(coords.x.intValue(), coords.y.intValue())`.

Hopefully knowing what we are actually using here helps us to make a better guess as to how and why that got slower in 752a8a8.

**Comment by sshock** [comment]
I think found the spot that caused the slowdown.  Take a close look at the changes to `safeHandle()` method in the `io.appium.uiautomator2.handler.Click` class in 752a8a8.

In the old code it would only call `Device.waitForIdle()` when doing the first block ("Click element command") but not when doing the second block ("tap command").  **After the change it calls Device.waitForIdle() in both cases.**

So..., it looks like my original hunch about waitForIdle() being the culprit was probably right after all.  But now I'm stuck because I already tried setting waitForIdleTimeout to 0 in a test earlier this week and although it definitely made the tapping go faster, it didn't work right because it only entered in like every other character.  But it makes no sense: why would it be safe and work ok for 3.8.1 to not do a waitForIdle() there but in 4.0.0 we have to pay that penalty or else?

**Comment by sshock** [comment]
I tried v4.0.0 with the patch below just now and sure enough, it made it go fast again, and it worked correctly too!

```patch
diff --git a/app/src/main/java/io/appium/uiautomator2/handler/Click.java b/app/src/main/java/io/appium/uiautomator2/handler/Click.java
index f341953..d3de2ab 100644
--- a/app/src/main/java/io/appium/uiautomator2/handler/Click.java
+++ b/app/src/main/java/io/appium/uiautomator2/handler/Click.java
@@ -1,5 +1,7 @@
 package io.appium.uiautomator2.handler;

+import android.util.Log;
+
 import org.json.JSONException;
 import org.json.JSONObject;

@@ -29,6 +31,7 @@ public class Click extends SafeRequestHandler {
             UiObjectNotFoundException {
         JSONObject payload = getPayload(request);
         if (payload.has(ELEMENT_ID_KEY_NAME)) {
+            Log.i("PLH", "HERE 1 CLICK");
             Logger.info("Click element command");
             String id = payload.getString(ELEMENT_ID_KEY_NAME);
             Session session = AppiumUIA2Driver.getInstance().getSessionOrThrow();
@@ -37,15 +40,18 @@ public class Click extends SafeRequestHandler {
                 return new AppiumResponse(getSessionId(request), WDStatus.NO_SUCH_ELEMENT);
             }
             element.click();
+            Log.i("PLH", "WAIT FOR IDLE");
+            Device.waitForIdle();
         } else {
+            Log.i("PLH", "HERE 1 TAP");
             Logger.info("tap command");
             Point coords = new Point(Double.parseDouble(payload.get("x").toString()),
                     Double.parseDouble(payload.get("y").toString()));
             coords = PositionHelper.getDeviceAbsPos(coords);
             final boolean res = getUiDevice().click(coords.x.intValue(), coords.y.intValue());
+            Log.i("PLH", "DON'T WAIT FOR IDLE");
             return new AppiumResponse(getSessionId(request), res);
         }
-        Device.waitForIdle();
         return new AppiumResponse(getSessionId(request), true);
     }
 }
```

But this begs the question of why I couldn't just set waitForIdleTimeout to 0 to work around this problem, but I'll test that again later today...

**Comment by mykola-mokhnach** [comment]
@sshock Please check if https://github.com/appium/appium-uiautomator2-server/pull/366 changes anything for you

**Comment by sshock** [comment]
@mykola-mokhnach cool, yeah I can set waitForIdle timeout to 0 again and try it both with and without your change there

I have to step out for a bit now, so it may be a few hours before I get a chance to test again...

**Comment by sshock** [comment]
@mykola-mokhnach good news!  it's working great now with waitForIdleTimeout 0, both with or without #366 ; I think it might even be slightly faster with #366 (runtimes were 58, 55, and 52 seconds) so that seems like a good change to keep.

I think earlier this week when I was testing I had set waitForIdleTimeout to 0 throughout the entire run, instead of just around my send_via_keyboard method (the one that calls single_tap a bunch of times).  This time I made sure to only set it to 0 just around that method.

So yeah, I think I'm good to go, thanks!  And yeah, let's keep #366 too.
